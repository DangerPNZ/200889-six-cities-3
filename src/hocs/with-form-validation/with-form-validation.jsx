import React from 'react';

const MIN_REVIEW_LENGTH = 50;

export const withFormValidation = (Component) => {
  return class WithFormValidation extends React.PureComponent {
    constructor(props) {
      super(props);
      this.state = {
        rating: ``,
        review: ``,
        notFilled: true
      };
      this.handleInputChange = this.handleInputChange.bind(this);
    }
    render() {
      return <Component
        {...this.props}
        onInputChange = {this.handleInputChange}
        notFilled = {this.state.notFilled}
        reviewData = {{
          comment: this.state.review,
          rating: Number(this.state.rating)
        }}
      />;
    }
    handleInputChange(event) {
      const {name, value} = event.target;
      this.setState({[name]: value}, this.checkValues);
    }
    checkValues() {
      this.setState({
        notFilled: (this.state.rating.length &&
          this.state.review.length &&
          this.state.review.length >= MIN_REVIEW_LENGTH) ? false : true
      });
    }
  };
};
