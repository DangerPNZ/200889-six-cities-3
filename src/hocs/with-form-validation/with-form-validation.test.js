import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import {withFormValidation} from './with-form-validation.jsx';

const TestComponent = ({notFilled, reviewData, onInputChange}) => (
  <div>
    {notFilled === true && <div>notFilled prop value equal to true</div>}
    {reviewData.comment === `` && <div>reviewData.comment is empty string</div>}
    {reviewData.rating === 0 && <div>reviewData.rating equal 0</div>}
    <textarea onChange={onInputChange}></textarea>
  </div>
);
const ResultComponent = withFormValidation(TestComponent);

it(`Test component, returned by HOC withFormValidation`, () => {
  const tree = renderer
  .create(
      <ResultComponent/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

TestComponent.propTypes = {
  reviewData: PropTypes.exact({
    comment: PropTypes.string,
    rating: PropTypes.number
  }),

  notFilled: PropTypes.bool.isRequired,

  onInputChange: PropTypes.func.isRequired
};

