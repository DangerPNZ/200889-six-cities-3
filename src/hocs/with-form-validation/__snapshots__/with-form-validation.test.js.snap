// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component, returned by HOC withFormValidation 1`] = `
<div>
  <div>
    notFilled prop value equal to true
  </div>
  <div>
    reviewData.comment is empty string
  </div>
  <div>
    reviewData.rating equal 0
  </div>
  <textarea
    onChange={[Function]}
  />
</div>
`;
